File: python/src/event_driven_client.py
"""
Web-Buddy Framework Event-Driven Python Client

Copyright (C) 2025-today  rydnr@acm-sl.org

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

import asyncio
import time
from typing import Dict, List, Optional, Any, Union, Type
from dataclasses import dataclass
from abc import ABC, abstractmethod
import aiohttp
import requests
from .events import (
    DomainEvent,
    EventResponse,
    ProjectSelectionRequested,
    ProjectSelected,
    ProjectSelectionFailed,
    ChatSelectionRequested,
    ChatSelected,
    ChatSelectionFailed,
    PromptSubmissionRequested,
    PromptSubmitted,
    PromptSubmissionFailed,
    ResponseRetrievalRequested,
    ResponseRetrieved,
    ResponseRetrievalFailed,
    GoogleImageDownloadRequested,
    GoogleImageDownloadCompleted,
    GoogleImageDownloadFailed,
    FileDownloadRequested,
    FileDownloadStarted,
    FileDownloadFailed,
    TrainingModeRequested,
    TrainingModeEnabled
)


@dataclass
class ClientConfig:
    """Configuration for Web-Buddy client"""
    base_url: str
    api_key: str
    timeout: int = 30
    retries: int = 3
    user_agent: str = 'WebBuddyPythonSDK/1.0.0'


class EventDrivenWebBuddyClient:
    """
    Event-Driven Web-Buddy Python Client
    
    Pure event-driven interface following the Web-Buddy Framework's EDA principles.
    All operations are performed by sending domain events and receiving event responses.
    """
    
    def __init__(self, config: ClientConfig):
        self.config = config
        self._session = None
        self._event_queue: Dict[str, asyncio.Future] = {}
    
    # === Core Event Sending Interface ===
    
    async def send_event(
        self,
        event: DomainEvent,
        extension_id: str,
        tab_id: int
    ) -> EventResponse:
        """
        Sends a domain event and waits for response.
        This is the low-level interface that all other methods use.
        """
        correlation_id = event.correlation_id or self._generate_correlation_id()
        
        try:
            # Create the dispatch payload in Web-Buddy format
            dispatch_payload = {
                'target': {
                    'extensionId': extension_id,
                    'tabId': tab_id
                },
                'message': {
                    'action': self._map_event_to_action(event),
                    'payload': self._extract_event_payload(event),
                    'correlationId': correlation_id
                }
            }
            
            # Send HTTP request to server
            response_data = await self._make_request('POST', '/api/dispatch', dispatch_payload)
            
            # Wait for and return the event response
            return await self._wait_for_event_response(correlation_id)
            
        except Exception as error:
            raise EventSendError(
                f'Failed to send event {event.__class__.__name__}: {str(error)}',
                event,
                error
            )
    
    async def send_events(
        self,
        events: List[Dict[str, Any]],
        parallel: bool = False,
        stop_on_error: bool = True
    ) -> List[EventResponse]:
        """
        Sends multiple events in sequence or parallel
        """
        if parallel:
            tasks = [
                self.send_event(event_data['event'], event_data['extension_id'], event_data['tab_id'])
                for event_data in events
            ]
            return await asyncio.gather(*tasks)
        else:
            results = []
            for event_data in events:
                try:
                    result = await self.send_event(
                        event_data['event'],
                        event_data['extension_id'],
                        event_data['tab_id']
                    )
                    results.append(result)
                except Exception as error:
                    if stop_on_error:
                        raise error
                    print(f"Warning: Event {event_data['event'].__class__.__name__} failed: {error}")
            return results
    
    # === High-Level Convenience Methods ===
    
    async def request_project_selection(
        self,
        extension_id: str,
        tab_id: int,
        project_name: str,
        selector: Optional[str] = None
    ) -> Union[ProjectSelected, ProjectSelectionFailed]:
        """Requests ChatGPT project selection"""
        event = ProjectSelectionRequested(
            project_name=project_name,
            selector=selector,
            correlation_id=self._generate_correlation_id()
        )
        
        return await self.send_event(event, extension_id, tab_id)
    
    async def request_chat_selection(
        self,
        extension_id: str,
        tab_id: int,
        chat_title: str,
        selector: Optional[str] = None
    ) -> Union[ChatSelected, ChatSelectionFailed]:
        """Requests chat selection"""
        event = ChatSelectionRequested(
            chat_title=chat_title,
            selector=selector,
            correlation_id=self._generate_correlation_id()
        )
        
        return await self.send_event(event, extension_id, tab_id)
    
    async def request_prompt_submission(
        self,
        extension_id: str,
        tab_id: int,
        prompt_text: str,
        selector: str = '#prompt-textarea'
    ) -> Union[PromptSubmitted, PromptSubmissionFailed]:
        """Requests prompt submission"""
        event = PromptSubmissionRequested(
            prompt_text=prompt_text,
            selector=selector,
            correlation_id=self._generate_correlation_id()
        )
        
        return await self.send_event(event, extension_id, tab_id)
    
    async def request_response_retrieval(
        self,
        extension_id: str,
        tab_id: int,
        selector: str = '[data-message-author-role="assistant"]',
        timeout: int = 30000
    ) -> Union[ResponseRetrieved, ResponseRetrievalFailed]:
        """Requests response retrieval from ChatGPT"""
        event = ResponseRetrievalRequested(
            selector=selector,
            timeout=timeout,
            correlation_id=self._generate_correlation_id()
        )
        
        return await self.send_event(event, extension_id, tab_id)
    
    async def request_google_image_download(
        self,
        extension_id: str,
        tab_id: int,
        image_element: Dict[str, Any],
        search_query: Optional[str] = None,
        filename: Optional[str] = None
    ) -> Union[GoogleImageDownloadCompleted, GoogleImageDownloadFailed]:
        """Requests Google Images download"""
        event = GoogleImageDownloadRequested(
            image_element=image_element,
            search_query=search_query,
            filename=filename,
            correlation_id=self._generate_correlation_id()
        )
        
        return await self.send_event(event, extension_id, tab_id)
    
    async def request_file_download(
        self,
        extension_id: str,
        tab_id: int,
        url: str,
        filename: Optional[str] = None,
        conflict_action: str = 'uniquify',
        save_as: bool = False
    ) -> Union[FileDownloadStarted, FileDownloadFailed]:
        """Requests file download"""
        event = FileDownloadRequested(
            url=url,
            filename=filename,
            conflict_action=conflict_action,
            save_as=save_as,
            correlation_id=self._generate_correlation_id()
        )
        
        return await self.send_event(event, extension_id, tab_id)
    
    async def request_training_mode(
        self,
        website: str
    ) -> TrainingModeEnabled:
        """Requests training mode activation"""
        event = TrainingModeRequested(
            website=website,
            correlation_id=self._generate_correlation_id()
        )
        
        # Training mode requests go directly to server
        return await self._send_training_event(event)
    
    # === Workflow Convenience Methods ===
    
    async def execute_full_chatgpt_workflow(
        self,
        extension_id: str,
        tab_id: int,
        workflow: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Complete ChatGPT workflow: select project, submit prompt, get response
        """
        results = {}
        
        # Step 1: Select project
        results['project_selection'] = await self.request_project_selection(
            extension_id, tab_id, workflow['project_name']
        )
        
        if isinstance(results['project_selection'], ProjectSelectionFailed):
            raise WorkflowError('Project selection failed', results)
        
        # Step 2: Select chat (optional)
        if workflow.get('chat_title'):
            results['chat_selection'] = await self.request_chat_selection(
                extension_id, tab_id, workflow['chat_title']
            )
            
            if isinstance(results['chat_selection'], ChatSelectionFailed):
                raise WorkflowError('Chat selection failed', results)
        
        # Step 3: Submit prompt
        results['prompt_submission'] = await self.request_prompt_submission(
            extension_id, tab_id, workflow['prompt_text']
        )
        
        if isinstance(results['prompt_submission'], PromptSubmissionFailed):
            raise WorkflowError('Prompt submission failed', results)
        
        # Step 4: Get response
        results['response_retrieval'] = await self.request_response_retrieval(
            extension_id, tab_id
        )
        
        return results
    
    async def download_multiple_google_images(
        self,
        extension_id: str,
        tab_id: int,
        images: List[Dict[str, Any]],
        parallel: bool = False,
        delay_between: float = 1.0
    ) -> List[Union[GoogleImageDownloadCompleted, GoogleImageDownloadFailed]]:
        """Batch Google Images download"""
        download_events = [
            {
                'event': GoogleImageDownloadRequested(
                    image_element=img['element'],
                    search_query=img.get('search_query'),
                    filename=img.get('filename', f'image_{i+1}'),
                    correlation_id=self._generate_correlation_id()
                ),
                'extension_id': extension_id,
                'tab_id': tab_id
            }
            for i, img in enumerate(images)
        ]
        
        if parallel:
            return await self.send_events(download_events, parallel=True)
        else:
            results = []
            for event_data in download_events:
                result = await self.send_event(
                    event_data['event'],
                    event_data['extension_id'],
                    event_data['tab_id']
                )
                results.append(result)
                
                if delay_between > 0:
                    await asyncio.sleep(delay_between)
                    
            return results
    
    # === Utility Methods ===
    
    async def ping(self) -> Dict[str, Any]:
        """Tests connectivity with server"""
        start_time = time.time()
        try:
            await self._make_request('GET', '/docs/health')
            return {
                'success': True,
                'latency': int((time.time() - start_time) * 1000)
            }
        except Exception:
            return {
                'success': False,
                'latency': int((time.time() - start_time) * 1000)
            }
    
    def get_config(self) -> ClientConfig:
        """Gets client configuration"""
        return self.config
    
    def update_config(self, updates: Dict[str, Any]) -> None:
        """Updates client configuration"""
        for key, value in updates.items():
            if hasattr(self.config, key):
                setattr(self.config, key, value)
    
    # === Private Helper Methods ===
    
    def _generate_correlation_id(self) -> str:
        """Generates unique correlation ID"""
        return f'py-client-{int(time.time() * 1000)}-{id(self) % 10000}'
    
    def _map_event_to_action(self, event: DomainEvent) -> str:
        """Maps domain event to Web-Buddy action"""
        event_type_map = {
            'ProjectSelectionRequested': 'SELECT_PROJECT',
            'ChatSelectionRequested': 'SELECT_CHAT',
            'PromptSubmissionRequested': 'FILL_PROMPT',
            'ResponseRetrievalRequested': 'GET_RESPONSE',
            'GoogleImageDownloadRequested': 'DOWNLOAD_IMAGE',
            'FileDownloadRequested': 'DOWNLOAD_FILE'
        }
        
        action = event_type_map.get(event.__class__.__name__)
        if not action:
            raise ValueError(f'No action mapping found for event type: {event.__class__.__name__}')
        return action
    
    def _extract_event_payload(self, event: DomainEvent) -> Dict[str, Any]:
        """Extracts payload from domain event"""
        if isinstance(event, ProjectSelectionRequested):
            return {
                'selector': event.selector or f'[data-project-name="{event.project_name}"]'
            }
        elif isinstance(event, PromptSubmissionRequested):
            return {
                'selector': event.selector,
                'value': event.prompt_text
            }
        elif isinstance(event, GoogleImageDownloadRequested):
            return {
                'selector': 'img',  # Will be refined by Google Images adapter
                'imageElement': event.image_element,
                'searchQuery': event.search_query,
                'filename': event.filename
            }
        elif isinstance(event, FileDownloadRequested):
            return {
                'url': event.url,
                'filename': event.filename,
                'conflictAction': event.conflict_action,
                'saveAs': event.save_as
            }
        
        # Default payload extraction
        return {
            key: value for key, value in event.__dict__.items()
            if key != 'correlation_id'
        }
    
    async def _make_request(self, method: str, endpoint: str, data: Optional[Dict] = None) -> Any:
        """Makes HTTP request with authentication and error handling"""
        url = f'{self.config.base_url}{endpoint}'
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {self.config.api_key}',
            'User-Agent': self.config.user_agent
        }
        
        if self._session is None:
            self._session = aiohttp.ClientSession()
        
        for attempt in range(self.config.retries):
            try:
                async with self._session.request(
                    method,
                    url,
                    headers=headers,
                    json=data,
                    timeout=aiohttp.ClientTimeout(total=self.config.timeout)
                ) as response:
                    if response.status >= 400:
                        error_data = await response.json() if response.content_type == 'application/json' else {}
                        raise EventSendError(f'HTTP {response.status}: {error_data.get("error", response.reason)}')
                    
                    return await response.json()
                    
            except asyncio.TimeoutError:
                if attempt == self.config.retries - 1:
                    raise
                await asyncio.sleep(2 ** attempt)  # Exponential backoff
            except Exception as error:
                if attempt == self.config.retries - 1:
                    raise
                await asyncio.sleep(2 ** attempt)
    
    async def _wait_for_event_response(self, correlation_id: str) -> EventResponse:
        """
        Waits for event response with matching correlation ID
        In a real implementation, this would listen for WebSocket responses
        or poll for the response. For now, we simulate immediate response.
        """
        # Placeholder - in real implementation would:
        # 1. Listen on WebSocket for responses with matching correlation_id
        # 2. Or poll a /api/responses/{correlation_id} endpoint
        # 3. Or use Server-Sent Events for real-time updates
        
        await asyncio.sleep(0.1)  # Simulate network delay
        
        # Return a mock successful response for now
        from .events import GenericEventResponse
        return GenericEventResponse(
            correlation_id=correlation_id,
            success=True,
            data={'message': 'Event processed successfully'}
        )
    
    async def _send_training_event(self, event: DomainEvent) -> EventResponse:
        """Sends training events to specific training endpoints"""
        endpoint = self._get_training_endpoint(event)
        payload = self._extract_event_payload(event)
        
        response_data = await self._make_request('POST', endpoint, payload)
        
        # Convert response to appropriate event response type
        from .events import TrainingModeEnabled
        return TrainingModeEnabled(
            session_id=response_data.get('sessionId', 'unknown'),
            website=response_data.get('website', ''),
            existing_patterns=response_data.get('existingPatterns', 0),
            correlation_id=event.correlation_id
        )
    
    def _get_training_endpoint(self, event: DomainEvent) -> str:
        """Gets appropriate training endpoint for event"""
        if isinstance(event, TrainingModeRequested):
            return '/api/training/enable'
        raise ValueError(f'No training endpoint for event: {event.__class__.__name__}')
    
    async def close(self):
        """Closes the client session"""
        if self._session:
            await self._session.close()
            self._session = None


# === Synchronous Client ===

class SyncEventDrivenWebBuddyClient:
    """
    Synchronous version of the event-driven client
    For use cases where async/await is not preferred
    """
    
    def __init__(self, config: ClientConfig):
        self.config = config
        self._session = requests.Session()
        self._session.headers.update({
            'Authorization': f'Bearer {config.api_key}',
            'Content-Type': 'application/json',
            'User-Agent': config.user_agent
        })
    
    def send_event(
        self,
        event: DomainEvent,
        extension_id: str,
        tab_id: int
    ) -> EventResponse:
        """Synchronous event sending"""
        correlation_id = event.correlation_id or self._generate_correlation_id()
        
        try:
            dispatch_payload = {
                'target': {'extensionId': extension_id, 'tabId': tab_id},
                'message': {
                    'action': self._map_event_to_action(event),
                    'payload': self._extract_event_payload(event),
                    'correlationId': correlation_id
                }
            }
            
            response_data = self._make_request('POST', '/api/dispatch', dispatch_payload)
            return self._create_mock_response(correlation_id)
            
        except Exception as error:
            raise EventSendError(
                f'Failed to send event {event.__class__.__name__}: {str(error)}',
                event,
                error
            )
    
    def request_project_selection(
        self,
        extension_id: str,
        tab_id: int,
        project_name: str,
        selector: Optional[str] = None
    ) -> Union[ProjectSelected, ProjectSelectionFailed]:
        """Synchronous project selection"""
        event = ProjectSelectionRequested(
            project_name=project_name,
            selector=selector,
            correlation_id=self._generate_correlation_id()
        )
        
        return self.send_event(event, extension_id, tab_id)
    
    def ping(self) -> Dict[str, Any]:
        """Synchronous connectivity test"""
        start_time = time.time()
        try:
            self._make_request('GET', '/docs/health')
            return {
                'success': True,
                'latency': int((time.time() - start_time) * 1000)
            }
        except Exception:
            return {
                'success': False,
                'latency': int((time.time() - start_time) * 1000)
            }
    
    def _make_request(self, method: str, endpoint: str, data: Optional[Dict] = None) -> Any:
        """Synchronous HTTP request"""
        url = f'{self.config.base_url}{endpoint}'
        
        try:
            if method == 'GET':
                response = self._session.get(url, timeout=self.config.timeout)
            elif method == 'POST':
                response = self._session.post(url, json=data, timeout=self.config.timeout)
            else:
                raise ValueError(f'Unsupported HTTP method: {method}')
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            raise EventSendError(f'Request failed: {str(e)}')
    
    def _generate_correlation_id(self) -> str:
        return f'py-sync-{int(time.time() * 1000)}-{id(self) % 10000}'
    
    def _map_event_to_action(self, event: DomainEvent) -> str:
        # Same as async version
        event_type_map = {
            'ProjectSelectionRequested': 'SELECT_PROJECT',
            'ChatSelectionRequested': 'SELECT_CHAT',
            'PromptSubmissionRequested': 'FILL_PROMPT',
            'ResponseRetrievalRequested': 'GET_RESPONSE',
            'GoogleImageDownloadRequested': 'DOWNLOAD_IMAGE',
            'FileDownloadRequested': 'DOWNLOAD_FILE'
        }
        
        action = event_type_map.get(event.__class__.__name__)
        if not action:
            raise ValueError(f'No action mapping found for event type: {event.__class__.__name__}')
        return action
    
    def _extract_event_payload(self, event: DomainEvent) -> Dict[str, Any]:
        # Same logic as async version
        if isinstance(event, ProjectSelectionRequested):
            return {
                'selector': event.selector or f'[data-project-name="{event.project_name}"]'
            }
        # ... other event types
        
        return {
            key: value for key, value in event.__dict__.items()
            if key != 'correlation_id'
        }
    
    def _create_mock_response(self, correlation_id: str) -> EventResponse:
        from .events import GenericEventResponse
        return GenericEventResponse(
            correlation_id=correlation_id,
            success=True,
            data={'message': 'Event processed successfully'}
        )


# === Custom Exceptions ===

class EventSendError(Exception):
    """Error occurred while sending event"""
    def __init__(self, message: str, event: DomainEvent = None, cause: Exception = None):
        super().__init__(message)
        self.event = event
        self.cause = cause


class WorkflowError(Exception):
    """Error occurred during workflow execution"""
    def __init__(self, message: str, partial_results: Dict[str, Any] = None):
        super().__init__(message)
        self.partial_results = partial_results or {}


# === Convenience Factory Functions ===

def create_async_client(base_url: str, api_key: str, **kwargs) -> EventDrivenWebBuddyClient:
    """Creates async event-driven client"""
    config = ClientConfig(base_url=base_url, api_key=api_key, **kwargs)
    return EventDrivenWebBuddyClient(config)


def create_sync_client(base_url: str, api_key: str, **kwargs) -> SyncEventDrivenWebBuddyClient:
    """Creates synchronous event-driven client"""
    config = ClientConfig(base_url=base_url, api_key=api_key, **kwargs)
    return SyncEventDrivenWebBuddyClient(config)

File: typescript/src/event-driven-client.js
"use strict";
/*
                        Web-Buddy Framework
                        Event-Driven TypeScript Client

    Copyright (C) 2025-today  rydnr@acm-sl.org

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowError = exports.EventSendError = exports.EventDrivenWebBuddyClient = void 0;
const types_1 = require("./types");
/**
 * Event-Driven Web-Buddy Client
 *
 * Pure event-driven interface following the Web-Buddy Framework's EDA principles.
 * All operations are performed by sending domain events and receiving event responses.
 */
class EventDrivenWebBuddyClient {
    constructor(config) {
        this.eventQueue = new Map();
        this.config = Object.assign({ timeout: 30000, retries: 3 }, config);
    }
    // === Core Event Sending Interface ===
    /**
     * Sends a domain event and waits for response
     * This is the low-level interface that all other methods use
     */
    sendEvent(event, extensionId, tabId) {
        return __awaiter(this, void 0, void 0, function* () {
            const correlationId = event.correlationId || this.generateCorrelationId();
            try {
                // Create the dispatch payload in Web-Buddy format
                const dispatchPayload = {
                    target: {
                        extensionId,
                        tabId
                    },
                    message: {
                        action: this.mapEventToAction(event),
                        payload: this.extractEventPayload(event),
                        correlationId
                    }
                };
                // Send HTTP request to server
                const response = yield this.makeRequest('POST', '/api/dispatch', dispatchPayload);
                // Wait for and return the event response
                return yield this.waitForEventResponse(correlationId);
            }
            catch (error) {
                throw new EventSendError(`Failed to send event ${event.constructor.name}: ${error.message}`, event, error);
            }
        });
    }
    /**
     * Sends multiple events in sequence
     */
    sendEvents(events, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { parallel = false, stopOnError = true } = options || {};
            if (parallel) {
                const promises = events.map(({ event, extensionId, tabId }) => this.sendEvent(event, extensionId, tabId));
                return yield Promise.all(promises);
            }
            else {
                const results = [];
                for (const { event, extensionId, tabId } of events) {
                    try {
                        const result = yield this.sendEvent(event, extensionId, tabId);
                        results.push(result);
                    }
                    catch (error) {
                        if (stopOnError) {
                            throw error;
                        }
                        // Continue with next event if not stopping on error
                        console.warn(`Event ${event.constructor.name} failed:`, error);
                    }
                }
                return results;
            }
        });
    }
    // === High-Level Convenience Methods ===
    // These provide a more user-friendly interface while maintaining event-driven architecture
    /**
     * Requests ChatGPT project selection
     */
    requestProjectSelection(extensionId, tabId, projectName, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.ProjectSelectionRequested(projectName, options === null || options === void 0 ? void 0 : options.selector, this.generateCorrelationId());
            return yield this.sendEvent(event, extensionId, tabId);
        });
    }
    /**
     * Requests chat selection
     */
    requestChatSelection(extensionId, tabId, chatTitle, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.ChatSelectionRequested(chatTitle, options === null || options === void 0 ? void 0 : options.selector, this.generateCorrelationId());
            return yield this.sendEvent(event, extensionId, tabId);
        });
    }
    /**
     * Requests prompt submission
     */
    requestPromptSubmission(extensionId, tabId, promptText, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.PromptSubmissionRequested(promptText, (options === null || options === void 0 ? void 0 : options.selector) || '#prompt-textarea', this.generateCorrelationId());
            return yield this.sendEvent(event, extensionId, tabId);
        });
    }
    /**
     * Requests response retrieval from ChatGPT
     */
    requestResponseRetrieval(extensionId, tabId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.ResponseRetrievalRequested((options === null || options === void 0 ? void 0 : options.selector) || '[data-message-author-role="assistant"]', (options === null || options === void 0 ? void 0 : options.timeout) || 30000, this.generateCorrelationId());
            return yield this.sendEvent(event, extensionId, tabId);
        });
    }
    /**
     * Requests file download
     */
    requestFileDownload(extensionId, tabId, url, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.FileDownloadRequested(url, options === null || options === void 0 ? void 0 : options.filename, options === null || options === void 0 ? void 0 : options.conflictAction, options === null || options === void 0 ? void 0 : options.saveAs, this.generateCorrelationId());
            return yield this.sendEvent(event, extensionId, tabId);
        });
    }
    /**
     * Requests training mode activation
     */
    requestTrainingMode(website) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.TrainingModeRequested(website, this.generateCorrelationId());
            // Training mode requests go directly to server, not through extension
            return yield this.sendTrainingEvent(event);
        });
    }
    /**
     * Requests automation pattern list
     */
    requestAutomationPatterns(filters) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = new types_1.AutomationPatternListRequested(filters, this.generateCorrelationId());
            return yield this.sendTrainingEvent(event);
        });
    }
    // === Workflow Convenience Methods ===
    /**
     * Complete ChatGPT workflow: select project, submit prompt, get response
     */
    executeFullChatGPTWorkflow(extensionId, tabId, workflow) {
        return __awaiter(this, void 0, void 0, function* () {
            const results = {};
            // Step 1: Select project
            results.projectSelection = yield this.requestProjectSelection(extensionId, tabId, workflow.projectName);
            if (results.projectSelection instanceof types_1.ProjectSelectionFailed) {
                throw new WorkflowError('Project selection failed', results);
            }
            // Step 2: Select chat (optional)
            if (workflow.chatTitle) {
                results.chatSelection = yield this.requestChatSelection(extensionId, tabId, workflow.chatTitle);
                if (results.chatSelection instanceof types_1.ChatSelectionFailed) {
                    throw new WorkflowError('Chat selection failed', results);
                }
            }
            // Step 3: Submit prompt
            results.promptSubmission = yield this.requestPromptSubmission(extensionId, tabId, workflow.promptText);
            if (results.promptSubmission instanceof types_1.PromptSubmissionFailed) {
                throw new WorkflowError('Prompt submission failed', results);
            }
            // Step 4: Get response
            results.responseRetrieval = yield this.requestResponseRetrieval(extensionId, tabId);
            return results;
        });
    }
    // === Utility Methods ===
    /**
     * Tests connectivity with a simple ping event
     */
    ping() {
        return __awaiter(this, void 0, void 0, function* () {
            const start = Date.now();
            try {
                yield this.makeRequest('GET', '/docs/health');
                return {
                    success: true,
                    latency: Date.now() - start
                };
            }
            catch (error) {
                return {
                    success: false,
                    latency: Date.now() - start
                };
            }
        });
    }
    /**
     * Gets client configuration
     */
    getConfig() {
        return Object.assign({}, this.config);
    }
    /**
     * Updates client configuration
     */
    updateConfig(updates) {
        this.config = Object.assign(Object.assign({}, this.config), updates);
    }
    // === Private Helper Methods ===
    generateCorrelationId() {
        return `client-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    }
    mapEventToAction(event) {
        const eventTypeMap = {
            'ProjectSelectionRequested': 'SELECT_PROJECT',
            'ChatSelectionRequested': 'SELECT_CHAT',
            'PromptSubmissionRequested': 'FILL_PROMPT',
            'ResponseRetrievalRequested': 'GET_RESPONSE',
            'FileDownloadRequested': 'DOWNLOAD_FILE'
        };
        const action = eventTypeMap[event.constructor.name];
        if (!action) {
            throw new Error(`No action mapping found for event type: ${event.constructor.name}`);
        }
        return action;
    }
    extractEventPayload(event) {
        // Extract payload based on event type
        if (event instanceof types_1.ProjectSelectionRequested) {
            return {
                selector: event.selector || `[data-project-name="${event.projectName}"]`
            };
        }
        if (event instanceof types_1.PromptSubmissionRequested) {
            return {
                selector: event.selector,
                value: event.promptText
            };
        }
        if (event instanceof types_1.FileDownloadRequested) {
            return {
                url: event.url,
                filename: event.filename,
                conflictAction: event.conflictAction,
                saveAs: event.saveAs
            };
        }
        // Default payload extraction
        return Object.fromEntries(Object.entries(event).filter(([key]) => key !== 'correlationId'));
    }
    makeRequest(method, endpoint, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = `${this.config.baseUrl}${endpoint}`;
            const headers = {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${this.config.apiKey}`,
                'User-Agent': this.config.userAgent || 'WebBuddyEventDrivenSDK/1.0.0'
            };
            const options = {
                method,
                headers,
                body: data ? JSON.stringify(data) : undefined
            };
            const response = yield this.fetchWithRetry(url, options);
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }
            return yield response.json();
        });
    }
    fetchWithRetry(url, options) {
        return __awaiter(this, void 0, void 0, function* () {
            let lastError;
            for (let attempt = 0; attempt < (this.config.retries || 3); attempt++) {
                try {
                    const controller = new AbortController();
                    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout || 30000);
                    const response = yield fetch(url, Object.assign(Object.assign({}, options), { signal: controller.signal }));
                    clearTimeout(timeoutId);
                    return response;
                }
                catch (error) {
                    lastError = error;
                    if (attempt < (this.config.retries || 3) - 1) {
                        yield this.delay(Math.pow(2, attempt) * 1000);
                    }
                }
            }
            throw lastError;
        });
    }
    waitForEventResponse(correlationId) {
        return __awaiter(this, void 0, void 0, function* () {
            // In a real implementation, this would listen for WebSocket responses
            // or poll for the response. For now, we'll simulate immediate response.
            // This is a placeholder - in the real implementation, this would:
            // 1. Listen on WebSocket for responses with matching correlationId
            // 2. Or poll a /api/responses/{correlationId} endpoint
            // 3. Or use Server-Sent Events for real-time updates
            return yield new Promise((resolve) => {
                setTimeout(() => {
                    resolve({
                        correlationId,
                        success: true,
                        timestamp: new Date()
                    });
                }, 100);
            });
        });
    }
    sendTrainingEvent(event) {
        return __awaiter(this, void 0, void 0, function* () {
            // Training events go to specific training endpoints
            const endpoint = this.getTrainingEndpoint(event);
            const payload = this.extractEventPayload(event);
            const response = yield this.makeRequest('POST', endpoint, payload);
            return response;
        });
    }
    getTrainingEndpoint(event) {
        if (event instanceof types_1.TrainingModeRequested) {
            return '/api/training/enable';
        }
        if (event instanceof types_1.AutomationPatternListRequested) {
            return '/api/training/patterns';
        }
        throw new Error(`No training endpoint for event: ${event.constructor.name}`);
    }
    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}
exports.EventDrivenWebBuddyClient = EventDrivenWebBuddyClient;
// === Custom Error Types ===
class EventSendError extends Error {
    constructor(message, event, cause) {
        super(message);
        this.event = event;
        this.cause = cause;
        this.name = 'EventSendError';
    }
}
exports.EventSendError = EventSendError;
class WorkflowError extends Error {
    constructor(message, partialResults) {
        super(message);
        this.partialResults = partialResults;
        this.name = 'WorkflowError';
    }
}
exports.WorkflowError = WorkflowError;
exports.default = EventDrivenWebBuddyClient;
